File: backend/routes/treatmentRoutes.js
const express = require("express");
const router = express.Router();

const db = require("../config/db");

// ==========================================
// ADD TREATMENT
// ==========================================

router.post("/", (req, res) => {
    const {
        patient_id,
        doctor_id,
        diagnosis,
        treatment_plan,
        medication,
        dosage,
        status,
        doctor_notes,
        follow_up_date
    } = req.body;

    if (
        !patient_id ||
        !doctor_id ||
        !diagnosis ||
        !treatment_plan
    ) {
        return res.status(400).json({
            success: false,
            message: "Patient, doctor, diagnosis and treatment plan are required"
        });
    }

    const sql = `
        INSERT INTO treatments
        (
            patient_id,
            doctor_id,
            diagnosis,
            treatment_plan,
            medication,
            dosage,
            status,
            doctor_notes,
            follow_up_date
        )
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
    `;

    db.query(
        sql,
        [
            patient_id,
            doctor_id,
            diagnosis,
            treatment_plan,
            medication || null,
            dosage || null,
            status || "Ongoing",
            doctor_notes || null,
            follow_up_date || null
        ],
        (err, result) => {

            if (err) {
                console.error(
                    "Treatment database error:",
                    err
                );

                return res.status(500).json({
                    success: false,
                    message: "Failed to save treatment",
                    error: err.message
                });
            }

            res.status(201).json({
                success: true,
                message: "Treatment saved successfully",
                treatment_id: result.insertId
            });
        }
    );
});

// ==========================================
// GET TREATMENTS FOR PATIENT
// ==========================================

router.get("/:patientId", (req, res) => {

    const sql = `
        SELECT *
        FROM treatments
        WHERE patient_id = ?
        ORDER BY created_at DESC
    `;

    db.query(
        sql,
        [req.params.patientId],
        (err, results) => {

            if (err) {
                console.error(
                    "Treatment fetch error:",
                    err
                );

                return res.status(500).json({
                    success: false,
                    message: "Failed to fetch treatments"
                });
            }

            res.json({
                success: true,
                treatments: results
            });
        }
    );
});

module.exports = router;

File: backend/routes/treatment_routes.py
from fastapi import APIRouter, Depends, HTTPException
from bson import ObjectId

from auth import require_roles

from shared import (
    TreatmentPlanInput,
    patients_collection,
    create_audit_log,
)

from datetime import datetime, timezone


router = APIRouter()


# =========================================================
# SAVE / UPDATE TREATMENT PLAN
# DOCTOR + SYSTEM ADMIN
# =========================================================


@router.patch("/api/patients/{patient_id}/treatment")
def update_treatment_plan(
    patient_id: str,
    treatment: TreatmentPlanInput,
    current_user: dict = Depends(
        require_roles(
            "Doctor",
            "System Administrator",
        )
    ),
):

    try:

        patient_object_id = ObjectId(patient_id)

    except Exception:

        raise HTTPException(
            status_code=400,
            detail="Invalid patient ID.",
        )

    if treatment.patient_id != patient_id:

        raise HTTPException(
            status_code=400,
            detail=(
                "Patient ID in request body does not "
                "match the URL."
            ),
        )

    patient = patients_collection.find_one(
        {
            "_id": patient_object_id
        }
    )

    if not patient:

        raise HTTPException(
            status_code=404,
            detail="Patient not found.",
        )

    # -----------------------------------------------------
    # DOCTOR CAN UPDATE ONLY ASSIGNED PATIENTS
    # -----------------------------------------------------

    if current_user["role"] == "Doctor":

        if patient.get("doctor_id") != current_user["id"]:

            raise HTTPException(
                status_code=403,
                detail=(
                    "You can only update treatment "
                    "for assigned patients."
                ),
            )

    # -----------------------------------------------------
    # VALIDATE FOLLOW-UP DATE
    # -----------------------------------------------------

    try:

        follow_up_date = datetime.strptime(
            treatment.follow_up_date,
            "%Y-%m-%d",
        ).date().isoformat()

    except ValueError:

        raise HTTPException(
            status_code=400,
            detail=(
                "Follow-up date must be in "
                "YYYY-MM-DD format."
            ),
        )

    # -----------------------------------------------------
    # CREATE TREATMENT PLAN
    # -----------------------------------------------------

    treatment_data = {
        "diagnosis": treatment.diagnosis.strip(),

        "medicines": treatment.medicines.strip(),

        "doctor_recommendations": (
            treatment.doctor_recommendations.strip()
        ),

        "follow_up_date": follow_up_date,

        "updated_at": datetime.now(timezone.utc),

        "updated_by": current_user["id"],

        "doctor_name": current_user.get(
            "name",
            "Doctor",
        ),
    }

    # -----------------------------------------------------
    # SAVE TO PATIENT DOCUMENT
    # -----------------------------------------------------

    patients_collection.update_one(
        {
            "_id": patient_object_id
        },
        {
            "$set": {
                "treatment_plan": treatment_data
            }
        },
    )
    create_audit_log(
        current_user,
        "UPDATE_TREATMENT_PLAN",
        resource=patient_id,
        details={
            "follow_up_date": follow_up_date,
            "diagnosis": treatment.diagnosis,
        },
    )

    return {
        "message": "Treatment plan saved successfully.",
        "patient_id": patient_id,
        "treatment_plan": treatment_data,
    }